File: python/processor/docspec_processor/extractor/privacy_extractor.py
"""Detect privacy / PII patterns in Python projects.

Targets:
  - ``@doc_pii`` decorator (DocSpec native)
  - ``@doc_sensitive`` decorator (DocSpec native)
  - Field name heuristics: ``email``, ``ssn``, ``phone``, ``address``, etc.
"""
# @docspec:module {
#   id: "docspec-py-privacy-extractor",
#   name: "Privacy Extractor",
#   description: "Detects PII fields via @doc_pii/@doc_sensitive decorators and field-name heuristics (email, SSN, phone, address, etc.), extracting privacy metadata with GDPR context.",
#   since: "3.0.0"
# }
from __future__ import annotations

import ast
import re
from typing import Any

from docspec_processor.extractor.extractor_interface import DocSpecExtractor, ExtractionContext

# Field name patterns that commonly hold PII
_PII_FIELD_PATTERNS: dict[str, str] = {
    r"(?i)email": "email",
    r"(?i)phone": "phone",
    r"(?i)ssn|social.?security": "government-id",
    r"(?i)passport": "government-id",
    r"(?i)tax.?id|tin": "government-id",
    r"(?i)credit.?card|card.?number": "financial",
    r"(?i)iban|account.?number": "financial",
    r"(?i)address|street|zip.?code|postal": "address",
    r"(?i)date.?of.?birth|dob|birth.?date": "date-of-birth",
    r"(?i)first.?name|last.?name|full.?name|surname": "name",
    r"(?i)ip.?addr": "ip-address",
    r"(?i)password|passwd|secret|token": "credential",
}


# @docspec:boundary "PII field detection and privacy metadata extraction"
class PrivacyExtractor(DocSpecExtractor):
    """Extracts privacy / PII metadata from decorators and field-name heuristics."""

    # @docspec:deterministic
    def extractor_name(self) -> str:
        return "privacy"

    # @docspec:intentional "Check if doc_pii/doc_sensitive decorators or common PII field names are present in the source"
    def is_available(self, context: ExtractionContext) -> bool:
        for item in context.sources:
            source: str = item["source"]
            if "doc_pii" in source or "doc_sensitive" in source:
                return True
            # Also enable if we find common PII field names in type-annotated classes
            if "email" in source.lower() or "password" in source.lower():
                return True
        return False

    # @docspec:intentional "Extract PII fields from doc_pii/doc_sensitive decorators and field-name heuristics"
    def extract(self, context: ExtractionContext) -> None:
        privacy_fields: list[dict[str, Any]] = []

        for item in context.sources:
            try:
                tree = ast.parse(item["source"])
            except SyntaxError:
                continue
            prefix = item["qualified_prefix"]

            for node in ast.walk(tree):
                if not isinstance(node, ast.ClassDef):
                    continue

                qualified = f"{prefix}.{node.name}"

                for stmt in node.body:
                    # --- Decorated fields via @doc_pii / @doc_sensitive ---
                    if isinstance(stmt, (ast.FunctionDef, ast.AsyncFunctionDef)):
                        pf = self._extract_from_decorator(stmt, qualified)
                        if pf:
                            privacy_fields.append(pf)

                    # --- Annotated assignments (class fields) ---
                    if isinstance(stmt, ast.AnnAssign) and isinstance(stmt.target, ast.Name):
                        field_name = stmt.target.id
                        pii_type = self._match_pii_pattern(field_name)
                        if pii_type:
                            privacy_fields.append({
                                "field": f"{qualified}.{field_name}",
                                "piiType": pii_type,
                                "detectedBy": "heuristic",
                            })

                    # --- Regular assignments ---
                    if isinstance(stmt, ast.Assign):
                        for target in stmt.targets:
                            if isinstance(target, ast.Name):
                                pii_type = self._match_pii_pattern(target.id)
                                if pii_type:
                                    privacy_fields.append({
                                        "field": f"{qualified}.{target.id}",
                                        "piiType": pii_type,
                                        "detectedBy": "heuristic",
                                    })

        if not privacy_fields:
            return

        existing: list[dict] = context.model.setdefault("privacy", [])
        # Deduplicate by field path
        existing_fields = {pf["field"] for pf in existing}
        for pf in privacy_fields:
            if pf["field"] not in existing_fields:
                existing.append(pf)
                existing_fields.add(pf["field"])

    # ------------------------------------------------------------------
    # Decorator extraction
    # ------------------------------------------------------------------

    # @docspec:intentional "Extract PII metadata from @doc_pii and @doc_sensitive decorator keyword arguments"
    def _extract_from_decorator(
        self, node: ast.FunctionDef | ast.AsyncFunctionDef, parent_qualified: str
    ) -> dict[str, Any] | None:
        """Look for ``@doc_pii(...)`` or ``@doc_sensitive`` on a method/property."""
        for dec in node.decorator_list:
            dec_name = self._decorator_name(dec)
            if dec_name == "doc_pii":
                pf: dict[str, Any] = {
                    "field": f"{parent_qualified}.{node.name}",
                    "detectedBy": "@doc_pii",
                }
                # Extract keyword args
                if isinstance(dec, ast.Call):
                    for kw in dec.keywords:
                        if kw.arg and isinstance(kw.value, ast.Constant):
                            if kw.arg == "value":
                                pf["piiType"] = kw.value.value
                            elif kw.arg == "retention":
                                pf["retention"] = kw.value.value
                            elif kw.arg == "gdpr_basis":
                                pf["gdprBasis"] = kw.value.value
                            elif kw.arg == "encrypted":
                                pf["encrypted"] = kw.value.value
                            elif kw.arg == "never_log":
                                pf["neverLog"] = kw.value.value
                            elif kw.arg == "never_return":
                                pf["neverReturn"] = kw.value.value
                    # First positional arg = pii type
                    if dec.args and isinstance(dec.args[0], ast.Constant):
                        pf.setdefault("piiType", dec.args[0].value)
                pf.setdefault("piiType", "other")
                return pf

            if dec_name == "doc_sensitive":
                return {
                    "field": f"{parent_qualified}.{node.name}",
                    "piiType": "other",
                    "neverLog": True,
                    "detectedBy": "@doc_sensitive",
                }
        return None

    # ------------------------------------------------------------------
    # Heuristic PII matching
    # ------------------------------------------------------------------

    # @docspec:deterministic
    @staticmethod
    def _match_pii_pattern(field_name: str) -> str | None:
        """Return PII type if *field_name* matches a known pattern, else ``None``."""
        for pattern, pii_type in _PII_FIELD_PATTERNS.items():
            if re.search(pattern, field_name):
                return pii_type
        return None

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    # @docspec:deterministic
    @staticmethod
    def _decorator_name(decorator: ast.expr) -> str | None:
        if isinstance(decorator, ast.Name):
            return decorator.id
        if isinstance(decorator, ast.Attribute):
            return decorator.attr
        if isinstance(decorator, ast.Call):
            func = decorator.func
            if isinstance(func, ast.Name):
                return func.id
            if isinstance(func, ast.Attribute):
                return func.attr
        return None

File: typescript/processor/src/extractor/privacy-extractor.ts
// @docspec:module {
//   id: "docspec-ts-privacy-extractor",
//   name: "Privacy Extractor",
//   description: "Detects PII fields via @DocPII/@DocSensitive decorators, class-validator PII decorators (@IsEmail, @IsPhoneNumber), and common PII field naming patterns (email, ssn, phone, etc.) to populate the privacy section of the DocSpec model.",
//   since: "3.0.0"
// }

import ts from "typescript";
import type {
  DocSpecExtractor,
  ExtractorContext,
  PrivacyFieldModel,
} from "./extractor-interface.js";

/**
 * Detects privacy-sensitive data annotations and patterns in TypeScript/Node.js
 * projects and populates the privacy section of the DocSpec model.
 *
 * @docspec:boundary "AST-based PII and privacy-sensitive field detection"
 *
 * Detection targets:
 * - DocPII decorator: `@DocPII('email')`, `@DocPII({ type: 'ssn', encrypted: true })`
 * - DocSensitive decorator: `@DocSensitive()`
 * - Common PII field naming patterns: email, ssn, phone, address, password, etc.
 * - class-validator decorators: `@IsEmail()`, `@IsPhoneNumber()`
 */
export class PrivacyExtractor implements DocSpecExtractor {
  /** Common field names that indicate PII data */
  private static readonly PII_FIELD_PATTERNS: Array<{ pattern: RegExp; piiType: string }> = [
    { pattern: /^email$/i, piiType: "email" },
    { pattern: /^e?mail(Address)?$/i, piiType: "email" },
    { pattern: /^(ssn|socialSecurityNumber|social_security)$/i, piiType: "ssn" },
    { pattern: /^phone(Number)?$/i, piiType: "phone" },
    { pattern: /^(mobile|cell)(Phone|Number)?$/i, piiType: "phone" },
    { pattern: /^(address|streetAddress|homeAddress)$/i, piiType: "address" },
    { pattern: /^(zip|zipCode|postalCode)$/i, piiType: "address" },
    { pattern: /^(password|passwd|pwd)$/i, piiType: "credential" },
    { pattern: /^(secret|token|apiKey|api_key)$/i, piiType: "credential" },
    { pattern: /^(firstName|lastName|fullName|name)$/i, piiType: "name" },
    { pattern: /^(dob|dateOfBirth|birthDate|birthday)$/i, piiType: "dob" },
    { pattern: /^(creditCard|cardNumber|ccNumber)$/i, piiType: "financial" },
    { pattern: /^(iban|bankAccount|accountNumber)$/i, piiType: "financial" },
    { pattern: /^(ip|ipAddress|ip_address)$/i, piiType: "ip-address" },
  ];

  /** class-validator decorators that imply PII */
  private static readonly VALIDATOR_PII_MAP: Record<string, string> = {
    IsEmail: "email",
    IsPhoneNumber: "phone",
    IsCreditCard: "financial",
    IsPassportNumber: "identity-document",
    IsISBN: "other",
  };

  /** @docspec:deterministic */
  extractorName(): string {
    return "privacy";
  }

  /** @docspec:deterministic */
  isAvailable(): boolean {
    return true;
  }

  /** @docspec:intentional "Scans all source files for PII-related decorators and naming patterns on class properties" */
  extract(context: ExtractorContext): void {
    const fields: PrivacyFieldModel[] = [];

    for (const sourceFile of context.sourceFiles) {
      if (sourceFile.isDeclarationFile || sourceFile.fileName.includes("node_modules")) continue;
      this.visitNode(sourceFile, sourceFile, fields);
    }

    if (fields.length > 0) {
      context.model.privacy.push(...fields);
    }
  }

  /** @docspec:intentional "Recursively walks the AST to detect property declarations with PII indicators" */
  private visitNode(
    node: ts.Node,
    sourceFile: ts.SourceFile,
    fields: PrivacyFieldModel[],
  ): void {
    // Check class properties for decorators and naming patterns
    if (ts.isPropertyDeclaration(node) && node.name) {
      this.analyzeProperty(node, sourceFile, fields);
    }

    ts.forEachChild(node, child => this.visitNode(child, sourceFile, fields));
  }

  /** @docspec:intentional "Checks a property for @DocPII, @DocSensitive, class-validator PII decorators, and PII field name patterns" */
  private analyzeProperty(
    node: ts.PropertyDeclaration,
    sourceFile: ts.SourceFile,
    fields: PrivacyFieldModel[],
  ): void {
    const fieldName = node.name.getText(sourceFile);
    const ownerClass = this.findEnclosingClassName(node);
    const qualifiedField = ownerClass ? `${ownerClass}.${fieldName}` : fieldName;

    // 1. Check for @DocPII decorator
    const decorators = ts.canHaveDecorators(node) ? ts.getDecorators(node) : undefined;
    if (decorators) {
      for (const decorator of decorators) {
        if (!ts.isCallExpression(decorator.expression)) continue;
        const decoratorName = decorator.expression.expression.getText(sourceFile);

        if (decoratorName === "DocPII") {
          const pf = this.parseDocPIIDecorator(decorator.expression, sourceFile, qualifiedField);
          fields.push(pf);
          return; // Already captured, skip further checks
        }

        if (decoratorName === "DocSensitive") {
          fields.push({
            field: qualifiedField,
            piiType: "other",
            neverLog: true,
          });
          return;
        }

        // class-validator decorators
        const piiType = PrivacyExtractor.VALIDATOR_PII_MAP[decoratorName];
        if (piiType) {
          fields.push({
            field: qualifiedField,
            piiType,
          });
          return;
        }
      }
    }

    // 2. Check field naming patterns (auto-detect PII from names)
    for (const { pattern, piiType } of PrivacyExtractor.PII_FIELD_PATTERNS) {
      if (pattern.test(fieldName)) {
        // Only flag if the field is in a class that looks like a model/entity/DTO
        if (this.isLikelyDataClass(node)) {
          fields.push({
            field: qualifiedField,
            piiType,
          });
        }
        return;
      }
    }
  }

  /** @docspec:deterministic */
  private parseDocPIIDecorator(
    callExpr: ts.CallExpression,
    sourceFile: ts.SourceFile,
    qualifiedField: string,
  ): PrivacyFieldModel {
    const pf: PrivacyFieldModel = { field: qualifiedField };

    if (callExpr.arguments.length === 0) {
      pf.piiType = "other";
      return pf;
    }

    const arg = callExpr.arguments[0];

    // @DocPII('email')
    if (ts.isStringLiteral(arg)) {
      pf.piiType = arg.text;
      return pf;
    }

    // @DocPII({ type: 'email', encrypted: true, retention: '90d', ... })
    if (ts.isObjectLiteralExpression(arg)) {
      for (const prop of arg.properties) {
        if (!ts.isPropertyAssignment(prop) || !ts.isIdentifier(prop.name)) continue;
        const key = prop.name.text;

        if (key === "type" && ts.isStringLiteral(prop.initializer)) {
          pf.piiType = prop.initializer.text;
        } else if (key === "value" && ts.isStringLiteral(prop.initializer)) {
          pf.piiType = prop.initializer.text;
        } else if (key === "retention" && ts.isStringLiteral(prop.initializer)) {
          pf.retention = prop.initializer.text;
        } else if (key === "gdprBasis" && ts.isStringLiteral(prop.initializer)) {
          pf.gdprBasis = prop.initializer.text;
        } else if (key === "encrypted") {
          pf.encrypted = prop.initializer.kind === ts.SyntaxKind.TrueKeyword;
        } else if (key === "neverLog") {
          pf.neverLog = prop.initializer.kind === ts.SyntaxKind.TrueKeyword;
        } else if (key === "neverReturn") {
          pf.neverReturn = prop.initializer.kind === ts.SyntaxKind.TrueKeyword;
        }
      }
    }

    if (!pf.piiType) pf.piiType = "other";
    return pf;
  }

  /**
   * Checks whether the containing class looks like a data transfer object,
   * entity, or model (to avoid false positives on controller fields, etc.).
   * @docspec:deterministic
   */
  private isLikelyDataClass(node: ts.Node): boolean {
    let current: ts.Node | undefined = node.parent;
    while (current) {
      if (ts.isClassDeclaration(current) && current.name) {
        const className = current.name.text.toLowerCase();
        return (
          className.includes("model") ||
          className.includes("entity") ||
          className.includes("dto") ||
          className.includes("user") ||
          className.includes("account") ||
          className.includes("profile") ||
          className.includes("customer") ||
          className.includes("patient") ||
          className.includes("person") ||
          className.includes("employee") ||
          className.includes("contact")
        );
      }
      current = current.parent;
    }
    return false;
  }

  /** @docspec:deterministic */
  private findEnclosingClassName(node: ts.Node): string | null {
    let current: ts.Node | undefined = node.parent;
    while (current) {
      if (ts.isClassDeclaration(current) && current.name) return current.name.text;
      current = current.parent;
    }
    return null;
  }
}
